librz/core/cmd/cmd_shell.c

```diff
@@ -586,3 +590,225 @@ RZ_IPI RzCmdStatus rz_cmd_shell_clippy_handler(RzCore *core, int argc, const cha
 	rz_cons_newline();
 	return RZ_CMD_STATUS_OK;
 }
+
+static char *showfile(char *res, const int nth, const char *fpath, const char *name, int printfmt) {
+#if __UNIX__
+	struct stat sb;
+#endif
+	const char *n = fpath;
+	char *nn, *u_rwx = NULL;
+	int sz = rz_file_size(n);
+	int perm, uid = 0, gid = 0;
+	int fch = '-';
+	if (!strncmp(fpath, "./", 2)) {
+		fpath = fpath + 2;
+	}
+	const bool isdir = rz_file_is_directory(n);
+	if (isdir) {
+		nn = rz_str_append(strdup(fpath), "/");
+	} else {
+		nn = strdup(fpath);
+	}
+	if (!*nn) {
+		free(nn);
+		return res;
+	}
+	perm = isdir ? 0755 : 0644;
+	if (!printfmt) {
+		needs_newline = ((nth + 1) % 4) ? 1 : 0;
+		res = rz_str_appendf(res, "%18s%s", nn, needs_newline ? "  " : "\n");
+		free(nn);
+		return res;
+	}
+	// TODO: escape non-printable chars in filenames
+	// TODO: Implement more real info in ls -l
+	// TODO: handle suid
+#if __UNIX__
+	if (lstat(n, &sb) != -1) {
+		ut32 ifmt = sb.st_mode & S_IFMT;
+		uid = sb.st_uid;
+		gid = sb.st_gid;
+		perm = sb.st_mode & 0777;
+		if (!(u_rwx = strdup(rz_str_rwx_i(perm >> 6)))) {
+			free(nn);
+			return res;
+		}
+		if (sb.st_mode & S_ISUID) {
+			u_rwx[2] = (sb.st_mode & S_IXUSR) ? 's' : 'S';
+		}
+		if (isdir) {
+			fch = 'd';
+		} else {
+			switch (ifmt) {
+			case S_IFCHR: fch = 'c'; break;
+			case S_IFBLK: fch = 'b'; break;
+			case S_IFLNK: fch = 'l'; break;
+			case S_IFIFO: fch = 'p'; break;
+#ifdef S_IFSOCK
+			case S_IFSOCK: fch = 's'; break;
+#endif
+			}
+		}
+	}
+#else
+	u_rwx = strdup("-");
+	fch = isdir ? 'd' : '-';
+#endif
+	if (printfmt == 'q') {
+		res = rz_str_appendf(res, "%s\n", nn);
+	} else if (printfmt == 'e') {
+		const char *eDIR = "📁";
+		const char *eIMG = "🌅";
+		const char *eHID = "👀";
+		const char *eANY = "  ";
+		// --
+		const char *icon = eANY;
+		if (isdir) {
+			icon = eDIR;
+#if __UNIX__
+		} else if ((sb.st_mode & S_IFMT) == S_IFLNK) {
+			const char *eLNK = "📎";
+			icon = eLNK;
+		} else if (sb.st_mode & S_ISUID) {
+			const char *eUID = "🔼";
+			icon = eUID;
+#endif
+		} else if (rz_str_casestr(nn, ".jpg") || rz_str_casestr(nn, ".png") || rz_str_casestr(nn, ".gif")) {
+			icon = eIMG;
+		} else if (*nn == '.') {
+			icon = eHID;
+		}
+		res = rz_str_appendf(res, "%s %s\n", icon, nn);
+	} else if (printfmt == FMT_RAW) {
+		res = rz_str_appendf(res, "%c%s%s%s  1 %4d:%-4d  %-10d  %s\n",
+			isdir ? 'd' : fch,
+			u_rwx ? u_rwx : "-",
+			rz_str_rwx_i((perm >> 3) & 7),
+			rz_str_rwx_i(perm & 7),
+			uid, gid, sz, nn);
+	} else if (printfmt == FMT_JSON) {
+		if (nth > 0) {
+			res = rz_str_append(res, ",");
+		}
+		res = rz_str_appendf(res, "{\"name\":\"%s\",\"size\":%d,\"uid\":%d,"
+					  "\"gid\":%d,\"perm\":%d,\"isdir\":%s}",
+			name, sz, uid, gid, perm, isdir ? "true" : "false");
+	}
+	free(nn);
+	free(u_rwx);
+	return res;
+}
+
+RZ_IPI RZ_OWN char *rz_syscmd_ls(RZ_NONNULL const int argc, const char **argv) {
```

Comment:
You can move this code directly into the handler's body

Reply:
Have you looked at line 775? Tell me how to handle that and then go to something like line 813 without refactoring the whole function? Or better yet, you do it

Reply:
Ok, makes sense, then you can keep it as a local static function then. No need for `RZ_IPI` even since it's only used in the `ls` handler which is itself IPI. I suggest to even rename it into `syscmd_ls`

Reply:
Done